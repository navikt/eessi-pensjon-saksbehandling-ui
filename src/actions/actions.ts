export interface Action<T> {
  type: string,
  payload?: T
}

export interface JustTypeAction {
  type: string
}

export interface PayloadError {
  error: any
}
