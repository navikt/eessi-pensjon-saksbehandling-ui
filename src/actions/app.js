import * as types from '../constants/actionTypes'
import * as urls from '../constants/urls'
import * as api from './api'

export function login () {
  let currentHost = window.location.origin // http://hostname
  let redirect = currentHost
  let context = encodeURIComponent(window.location.pathname + window.location.search)

  let newUrl = urls.LOGIN_URL + '?redirect=' + redirect + '&context=' + context
  window.location.href = newUrl

  return {
    type: types.APP_LOGIN_REQUEST
  }
}

export function logout () {
  let redirectUrl = urls.LOGOUT_URL
  window.location.href = redirectUrl
  return {
    type: types.APP_LOGOUT_REQUEST
  }
}

export function getUserInfo () {
  return api.call({
    url: urls.API_USERINFO_URL,
    type: {
      request: types.APP_USERINFO_REQUEST,
      success: types.APP_USERINFO_SUCCESS,
      failure: types.APP_USERINFO_FAILURE
    }
  })
}

export function getAndPrefillPersonName () {
  return function (dispatch) {
    dispatch({ type: types.APP_GET_AND_PREFILL_PERSON_NAME })
    try {
      dispatch(getPersonData())
      dispatch(suggestPersonNameFromUsernameIfNotInState())
    } catch (error) {
      dispatch({ type: types.APP_GET_AND_PREFILL_PERSON_NAME_FAILURE, payload: error })
    }
  }
}

export function getPersonData () {
  return api.call({
    url: urls.API_PERSONDATA_URL,
    type: {
      request: types.APP_PERSONDATA_REQUEST,
      success: types.APP_PERSONDATA_SUCCESS,
      failure: types.APP_PERSONDATA_FAILURE
    }
  })
}

export function suggestPersonNameFromUsernameIfNotInState () {
  return function (dispatch, globalState) {
    if (!globalState.pinfo.person.nameAtBirth) {
      let lastName = globalState.app.lastName
      return dispatch({
        type: types.PINFO_PERSON_SET,
        payload: {
          nameAtBirth: lastName
        }
      })
    }
  }
}

export function clearData () {
  return {
    type: types.APP_CLEAR_DATA
  }
}

export function setReferrer (referrer) {
  return {
    type: types.APP_REFERRER_SET,
    payload: {
      referrer: referrer
    }
  }
}

export function registerDroppable (id, ref) {
  return {
    type: types.APP_DROPPABLE_REGISTER,
    payload: {
      id: id,
      ref: ref
    }
  }
}

export function unregisterDroppable (id) {
  return {
    type: types.APP_DROPPABLE_UNREGISTER,
    payload: {
      id: id
    }
  }
}
