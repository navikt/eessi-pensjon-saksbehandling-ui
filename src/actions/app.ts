import * as types from 'constants/actionTypes'
import * as urls from 'constants/urls'
import * as api from 'eessi-pensjon-ui/dist/api'
import { ActionWithPayload, ThunkResult } from 'eessi-pensjon-ui/dist/declarations/types'
import { Action, ActionCreator } from 'redux'
import samplePerson from 'resources/tests/samplePerson'

const sprintf = require('sprintf-js').sprintf

export interface ParamPayload {
  key: string,
  value?: any
}

export const setStatusParam: ActionCreator<ActionWithPayload> = (key: string, value: any): ActionWithPayload<ParamPayload> => ({
  type: types.APP_PARAM_SET,
  payload: {
    key: key,
    value: value
  }
})

export const unsetStatusParam: ActionCreator<ActionWithPayload> = (key: string): ActionWithPayload<ParamPayload> => ({
  type: types.APP_PARAM_UNSET,
  payload: {
    key: key
  }
})

export const login = (): Action => {
  const redirect = window.location.origin // http://hostname
  const context = encodeURIComponent(window.location.pathname + window.location.search)
  window.location.href = urls.LOGIN_URL + '?redirect=' + redirect + '&context=' + context
  return {
    type: types.APP_LOGIN_REQUEST
  }
}

export const logout = (): Action => {
  window.location.href = urls.LOGOUT_URL
  return {
    type: types.APP_LOGOUT_REQUEST
  }
}

export const getUserInfo: ActionCreator<ThunkResult<ActionWithPayload>> = (): ThunkResult<ActionWithPayload> => {
  return api.call({
    url: urls.API_USERINFO_URL,
    cascadeFailureError: true,
    expectedPayload: {
      subject: 'demoSaksbehandlerUser',
      role: 'SAKSBEHANDLER',
      allowed: true,
      features: {
        P5000_VISIBLE: true
      }
    },
    type: {
      request: types.APP_USERINFO_REQUEST,
      success: types.APP_USERINFO_SUCCESS,
      failure: types.APP_USERINFO_FAILURE,
      forbidden: types.APP_USERINFO_FORBIDDEN
    }
  })
}

export const getPersonInfo: ActionCreator<ThunkResult<ActionWithPayload>> = (aktoerId: string): ThunkResult<ActionWithPayload> => {
  return api.call({
    url: sprintf(urls.PERSON_URL, { aktoerId: aktoerId }),
    expectedPayload: samplePerson,
    type: {
      request: types.APP_PERSONINFO_REQUEST,
      success: types.APP_PERSONINFO_SUCCESS,
      failure: types.APP_PERSONINFO_FAILURE
    }
  })
}

export const clearData = (): Action => ({
  type: types.APP_CLEAR_DATA
})
