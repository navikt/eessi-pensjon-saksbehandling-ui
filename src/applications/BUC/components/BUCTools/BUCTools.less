.a-buc-c-buctools {
  opacity: 0;
  transform: translateX(20px);
  animation: slideInFromRight 0.3s forwards;

  &__comment-textarea {
    min-height: 150px;
    width: 100%;
  }
}
