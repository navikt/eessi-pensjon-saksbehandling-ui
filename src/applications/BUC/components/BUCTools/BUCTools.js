import React, { useState, useEffect } from 'react'
import PT from 'prop-types'
import { connect, bindActionCreators } from 'store'
import classNames from 'classnames'
import Icons from 'components/ui/Icons'
import { EkspanderbartpanelBase, Ingress, Normaltekst, Textarea, Flatknapp } from 'components/ui/Nav'
import MultipleSelect from 'components/ui/MultipleSelect/MultipleSelect'
import * as bucActions from 'actions/buc'

const mapStateToProps = (state) => {
  return {
    tagList: state.buc.tagList,
    bucsInfo: state.buc.bucsInfo,
    gettingTagList: state.loading.gettingTagList,
    aktoerId: state.app.params.aktoerId,
    savingBucsInfo: state.loading.savingBucsInfo
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    actions: bindActionCreators(bucActions, dispatch)
  }
}

const BUCTools = (props) => {

  const { t, aktoerId, buc, bucInfo, bucsInfo, savingBucsInfo, tagList, gettingTagList, actions, locale, className } = props
  const [ tags, setTags ] = useState(bucInfo && bucInfo.tags ? bucInfo.tags.map(tag => {
    return {
      value: tag,
      label: t('buc:tag-' + tag)
    }
  }) : [])
  const [ comment, setComment ] = useState(bucInfo ? bucInfo.comment : undefined)
  const [ allTags, setAllTags ] = useState(undefined)

  useEffect(() => {
    if (tagList === undefined && !gettingTagList) {
      actions.getTagList()
    }
  }, [actions, gettingTagList, tagList])

  useEffect(() => {
    if (!allTags && tagList) {
      setAllTags(tagList.map(tag => {
        return {
          value: tag,
          label: t('buc:tag-' + tag)
        }
      }))
    }
  }, [t, allTags, tagList])

  const onTagsChange = (tagsList) => {
    setTags(tagsList)
  }

  const onCommentChange = (e) => {
    setComment(e.target.value)
  }

  const onSaveButtonClick = () => {
    actions.saveBucsInfo({
      bucsInfo: bucsInfo,
      aktoerId: aktoerId,
      tags: tags,
      comment: comment,
      buc: buc
    })
  }

  const renderTags = () => {
    return <div className='flex-fill'>
      <div className='a-buc-c-buctools__form-tags-label d-flex mb-2 align-items-center'>
        <Icons kind='tilsette' />
        <span className='ml-2'>{t('buc:form-tagsForBUC')}</span>
      </div>
      <div className='mb-3'>
        <Normaltekst className='mb-2'>{t('buc:form-tagsForBUC-description')}</Normaltekst>
        <MultipleSelect
          className='a-buc-c-buctools__tags flex-fill'
          id='a-buc-c-buctools__tags-select-id'
          placeholder={t('buc:form-tagPlaceholder')}
          aria-describedby='help-tags'
          locale={locale}
          values={tags}
          hideSelectedOptions={false}
          onChange={onTagsChange}
          optionList={allTags} />
      </div>
    </div>
  }

  const renderComments = () => {
    return <div className='flex-fill'>
      <div className='a-buc-c-buctools__form-tags-label d-flex mb-2 align-items-center'>
        <Icons kind='tilsette' />
        <span className='ml-2'>{t('buc:form-commentForBUC')}</span>
      </div>
      <Textarea
        label={''}
        value={comment}
        style={{ minHeight: '150px' }}
        onChange={onCommentChange} />
    </div>
  }

  return <EkspanderbartpanelBase
    className={classNames('a-buc-c-buctools', className)}
    id='a-buc-c-buctools__panel-id'
    heading={<Ingress
      className='a-buc-c-buctoold__title'>
      {t('buc:form-BUCtools')}
    </Ingress>
    }>
    {renderTags()}
    {renderComments()}
    <Flatknapp
      id='a-buc-c-buctools__form-comments-button'
      className='a-buc-c-buctools__form-button'
      disabled={savingBucsInfo}
      onClick={onSaveButtonClick}>{savingBucsInfo ? t('ui:saving'): t('ui:change')}</Flatknapp>
  </EkspanderbartpanelBase>
}

BUCTools.propTypes = {
  t: PT.func.isRequired,
  buc: PT.object.isRequired,
  className: PT.string
}

export default connect(mapStateToProps, mapDispatchToProps)(BUCTools)
