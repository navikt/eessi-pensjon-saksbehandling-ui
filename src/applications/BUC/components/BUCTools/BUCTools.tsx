import { getTagList, saveBucsInfo } from 'actions/buc'
import { sedFilter } from 'applications/BUC/components/BUCUtils/BUCUtils'
import SEDP5000 from 'applications/SEDP5000/SEDP5000'
import { BUCMode } from 'applications/BUC/index'
import Trashcan from 'assets/icons/Trashcan'
import MultipleSelect from 'components/MultipleSelect/MultipleSelect'
import { AllowedLocaleString, FeatureToggles, Loading } from 'declarations/app.d'

import {
  Buc,
  BucInfo,
  BucsInfo,
  Comment,
  Comments,
  SedContentMap,
  Tag,
  TagRawList,
  Tags,
  ValidBuc
} from 'declarations/buc'
import { BucInfoPropType, BucPropType } from 'declarations/buc.pt'
import { State } from 'declarations/reducers'
import _ from 'lodash'
import { buttonLogger, standardLogger } from 'metrics/loggers'
import { HoyreChevron } from 'nav-frontend-chevron'
import { Element, Normaltekst, Undertittel } from 'nav-frontend-typografi'
import NavHighContrast, {
  HighContrastKnapp,
  HighContrastPanel,
  HighContrastTabs,
  HighContrastTextArea,
  HorizontalSeparatorDiv,
  slideInFromRight,
  theme,
  themeHighContrast,
  themeKeys,
  VerticalSeparatorDiv
} from 'nav-hoykontrast'

import PT from 'prop-types'
import { useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { useDispatch, useSelector } from 'react-redux'
import { ValueType } from 'react-select'
import styled from 'styled-components'

const BUCToolsPanel = styled(HighContrastPanel)`
  opacity: 0;
  transform: translateX(20px);
  animation: ${slideInFromRight(20)} 0.3s forwards;
  &.loading {
    background-color: rgba(128,128,128,0.2);
  }
`
const CommentDiv = styled.div`
  border-bottom-width: ${({ theme }) => theme.type === 'themeHighContrast' ? '2px' : '1px'};
  border-bottom-style: solid;
  border-bottom-color: ${({ theme }) => theme[themeKeys.MAIN_BORDER_COLOR]};
  margin-top: 0.5rem;
  margin-bottom: 0.5rem;
  display: flex;
  justify-content: space-between;
`
const FlexDiv = styled.div`
  display: flex;

`
const P5000Div = styled.div`
  position: relative;
  margin-bottom: 1rem;
`
const PaddedTabContent = styled.div`
  padding-top: 1rem;
  padding-bottom: 1rem;
`
const RemoveComment = styled.div`
  cursor: pointer;
`
export const TextArea = styled(HighContrastTextArea)`
  min-height: 150px;
  width: 100%;
`

export interface BUCToolsProps {
  aktoerId: string
  buc: Buc
  bucInfo: BucInfo
  className?: string
  initialTab?: number
  onTagChange?: (tagList: Tags) => void
  setMode: (mode: BUCMode, s: string, callback?: () => void, content?: JSX.Element) => void
}

export interface BUCToolsSelector {
  bucsInfo?: BucsInfo | undefined
  featureToggles: FeatureToggles
  highContrast: boolean
  loading: Loading
  locale: AllowedLocaleString
  sedContent: SedContentMap
  tagList?: TagRawList | undefined
}

const mapState = (state: State): BUCToolsSelector => ({
  bucsInfo: state.buc.bucsInfo,
  featureToggles: state.app.featureToggles,
  highContrast: state.ui.highContrast,
  loading: state.loading,
  locale: state.ui.locale,
  sedContent: state.buc.sedContent,
  tagList: state.buc.tagList
})

const BUCTools: React.FC<BUCToolsProps> = ({
  aktoerId,
  buc,
  bucInfo,
  initialTab = 0,
  onTagChange,
  setMode
}: BUCToolsProps): JSX.Element => {
  const {
    bucsInfo, featureToggles, highContrast, loading, tagList
  }: BUCToolsSelector = useSelector<State, BUCToolsSelector>(mapState)
  const dispatch = useDispatch()
  const { t } = useTranslation()

  const [_activeTab, setActiveTab] = useState<number>(initialTab)
  const [_allTags, setAllTags] = useState<Tags | undefined>(undefined)
  const [_comment, setComment] = useState< string | null | undefined >('')

  const [_originalComments, setOriginalComments] = useState<Comments | string | null | undefined >(bucInfo ? bucInfo.comment : '')

  const [_tags, setTags] = useState<Tags | undefined>(undefined)
  const _theme = highContrast ? themeHighContrast : theme

  const onTagsChange = (tagsList: ValueType<Tag, true>): void => {
    if (tagsList) {
      if (_.isFunction(onTagChange)) {
        onTagChange(tagsList as Tags)
      }
      standardLogger('buc.edit.tools.tags.select', { tags: (tagsList as Tags)?.map(t => t.label) || [] })
      setTags(tagsList as Tags)
      dispatch(saveBucsInfo({
        bucsInfo: bucsInfo!,
        aktoerId: aktoerId,
        tags: _tags ? (_tags as Tags).map(tag => tag.value) : [],
        comment: _originalComments,
        buc: buc as ValidBuc
      }))
    }
  }

  const onCommentChange = (e: React.ChangeEvent<HTMLTextAreaElement>): void => {
    setComment(e.target.value)
  }

  const onSaveCommentClick = (): void => {
    standardLogger('buc.edit.tools.comment.textarea', { comment: _comment })
    let newOriginalComments: Comments | string = _originalComments ? _.cloneDeep(_originalComments) : []
    if (_.isString(newOriginalComments)) {
      newOriginalComments = [{ value: newOriginalComments }]
    }
    const newComment: Comment = {
      value: _comment!
    }
    if (_comment) {
      newOriginalComments!.push(newComment)
      setOriginalComments(newOriginalComments)
      setComment('')
    }
    dispatch(saveBucsInfo({
      bucsInfo: bucsInfo!,
      aktoerId: aktoerId,
      tags: _tags ? _tags.map(tag => tag.value) : [],
      comment: newOriginalComments,
      buc: buc as ValidBuc
    }))
  }

  const onDeleteComment = (i: number): void => {
    if (window.confirm(t('buc:form-areYouSureDeleteComment'))) {
      const newOriginalComments: Comments = _.cloneDeep(_originalComments) as Comments
      newOriginalComments.splice(i, 1)
      setOriginalComments(newOriginalComments)
      dispatch(saveBucsInfo({
        bucsInfo: bucsInfo!,
        aktoerId: aktoerId,
        tags: _tags ? _tags.map(tag => tag.value) : [],
        comment: newOriginalComments,
        buc: buc as ValidBuc
      }))
    }
  }

  const onGettingP5000Click = (e: React.MouseEvent): void => {
    buttonLogger(e)
    setMode('p5000', 'forward', undefined, (
      <SEDP5000
        buc={buc}
        setMode={setMode}
        context='overview'
      />
    ))
  }

  const tabs = [{
    label: t('buc:form-labelP5000'),
    key: 'P5000'
  }, {
    label: t('ui:tags'),
    key: 'tags'
  }, {
    label: t('buc:form-commentForBUC'),
    key: 'comments'
  }]

  const hasP5000s = (): boolean => (!_.isEmpty(buc?.seds?.filter(sedFilter).filter(sed => sed.type === 'P5000' && sed.status !== 'cancelled')))

  useEffect(() => {
    if (tagList === undefined && !loading.gettingTagList) {
      dispatch(getTagList())
    }
  }, [dispatch, loading, tagList])

  useEffect(() => {
    if (!_allTags && tagList) {
      setAllTags(tagList.map((tag: string) => ({
        value: tag,
        label: t('buc:' + tag)
      })))
    }
  }, [t, _allTags, tagList])

  useEffect(() => {
    if (bucInfo && bucInfo.tags && _tags === undefined) {
      setTags(bucInfo.tags.map((tag: string) => ({
        value: tag,
        label: t('buc:' + tag)
      })))
    }
  }, [bucInfo, t, _tags])

  return (
    <NavHighContrast highContrast={highContrast}>
      <BUCToolsPanel
        data-test-id='a-buc-c-buctools__panel-id'
      >
        <>
          <HighContrastTabs
            data-test-id='a-buc-c-buctools__tabs-id'
            onChange={(e: any, i: number) => setActiveTab(i)}
            tabs={tabs}
            defaultAktiv={_activeTab}
          />
          <PaddedTabContent>
            {tabs[_activeTab].key === 'P5000' && (
              <P5000Div>
                <Undertittel>
                  {t('buc:form-titleP5000')}
                </Undertittel>
                <VerticalSeparatorDiv />
                <FlexDiv>
                  <HighContrastKnapp
                    data-amplitude='buc.edit.tools.P5000.view'
                    data-test-id='a-buc-c-buctools__P5000-button-id'
                    disabled={!hasP5000s()}
                    onClick={onGettingP5000Click}
                  >
                    {featureToggles.P5000_SUMMER_VISIBLE ? t('buc:form-seeP5000s') : t('buc:form-viewP5000s')}

                    <HorizontalSeparatorDiv size='0.3' />
                    <HoyreChevron />
                  </HighContrastKnapp>
                </FlexDiv>
              </P5000Div>
            )}
            {tabs[_activeTab].key === 'tags' && (
              <>
                <VerticalSeparatorDiv size='0.5' />
                <Normaltekst>
                  {t('buc:form-tagsForBUC-description')}
                </Normaltekst>
                <VerticalSeparatorDiv size='0.5' />
                {_tags && !_.isEmpty(_tags) && (
                  <>
                    <dt>
                      <Element>
                        {t('buc:form-tagsForBUC')}:
                      </Element>
                    </dt>
                    <dd>
                      <Normaltekst>
                        {_tags.map((tag: Tag) => tag.label).join(', ')}
                      </Normaltekst>
                    </dd>
                  </>
                )}
                <VerticalSeparatorDiv size='0.5' />
                <MultipleSelect<Tag>
                  ariaLabel={t('buc:form-tagsForBUC')}
                  aria-describedby='help-tags'
                  data-test-id='a-buc-c-buctools__tags-select-id'
                  hideSelectedOptions={false}
                  highContrast={highContrast}
                  label=''
                  onSelect={onTagsChange}
                  options={_allTags}
                  placeholder={t('buc:form-tagPlaceholder')}
                  values={_tags || []}
                />
              </>
            )}
            {tabs[_activeTab].key === 'comments' && (
              <>
                <VerticalSeparatorDiv size='0.5' />
                <Element>
                  {t('ui:comment')}
                </Element>
                {_originalComments
                  ? (_originalComments as Comments).map((comment: Comment, i: number) => (
                    <CommentDiv
                      data-test-id='a-buc-c-buctools__comment-div-id'
                      key={i}
                    >
                      <Normaltekst>
                        {comment.value}
                      </Normaltekst>
                      <RemoveComment>
                        <Trashcan
                          data-test-id={'a-buc-c-buctools__comment-delete-' + i + '-id'}
                          width={20}
                          color={_theme[themeKeys.MAIN_INTERACTIVE_COLOR]}
                          onClick={() => onDeleteComment(i)}
                        />
                      </RemoveComment>
                    </CommentDiv>
                    ))
                  : (
                    <Normaltekst>
                      {t('ui:noCommentsYet')}
                    </Normaltekst>
                    )}
                <VerticalSeparatorDiv size='0.5' />
                <TextArea
                  data-test-id='a-buc-c-buctools__comment-textarea-id'
                  className='skjemaelement__input'
                  label=''
                  value={_comment || ''}
                  onChange={onCommentChange}
                />
                <HighContrastKnapp
                  data-test-id='a-buc-c-buctools__comment-save-button-id'
                  disabled={loading.savingBucsInfo}
                  spinner={loading.savingBucsInfo}
                  onClick={onSaveCommentClick}
                >
                  {loading.savingBucsInfo ? t('ui:saving') : t('ui:add')}
                </HighContrastKnapp>
              </>
            )}

          </PaddedTabContent>
        </>
      </BUCToolsPanel>
    </NavHighContrast>
  )
}

BUCTools.propTypes = {
  aktoerId: PT.string.isRequired,
  buc: BucPropType.isRequired,
  bucInfo: BucInfoPropType.isRequired,
  className: PT.string,
  initialTab: PT.number,
  onTagChange: PT.func
}

export default BUCTools
