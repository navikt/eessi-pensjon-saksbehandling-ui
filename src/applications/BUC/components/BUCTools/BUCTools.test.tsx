import { getTagList, saveBucsInfo } from 'actions/buc'
import { Buc, BucInfo, BucsInfo, Comment, Comments, P5000FromRinaMap, TagRawList } from 'declarations/buc.d'
import { AllowedLocaleString } from 'declarations/app.d'
import { mount, ReactWrapper } from 'enzyme'
import mockFeatureToggles from 'mocks/app/featureToggles'
import mockBucs from 'mocks/buc/bucs'
import mockTagList from 'mocks/buc/tagsList'
import mockBucsInfo from 'mocks/buc/bucsInfo'
import mockP50001 from 'mocks/buc/sed_P5000_small1'
import mockP50002 from 'mocks/buc/sed_P5000_small2'
import { stageSelector } from 'setupTests'
import BUCTools, { BUCToolsProps, TextArea } from './BUCTools'
import allTags from 'constants/tagsList'

jest.mock('actions/buc', () => ({
  getSed: jest.fn(),
  getTagList: jest.fn(),
  saveBucsInfo: jest.fn()
}))

const defaultSelector = {
  bucsInfo: mockBucsInfo as BucsInfo,
  featureToggles: mockFeatureToggles,
  highContrast: false,
  loading: {},
  locale: 'nb' as AllowedLocaleString,
  p5000FromRinaMap: {
    '60578cf8bf9f45a7819a39987c6c8fd4': mockP50001,
    '50578cf8bf9f45a7819a39987c6c8fd4': mockP50002
  } as P5000FromRinaMap,
  tagList: mockTagList as TagRawList
}

const buc: Buc = mockBucs()[0]
const bucInfo: BucInfo = (mockBucsInfo as BucsInfo).bucs['' + buc.caseId]

describe('applications/BUC/components/BUCTools/BUCTools', () => {
  let wrapper: ReactWrapper

  const initialMockProps: BUCToolsProps = {
    aktoerId: '123',
    buc: buc,
    bucInfo: bucInfo,
    initialTab: 0,
    onTagChange: jest.fn(),
    setMode: jest.fn()
  } as BUCToolsProps

  beforeEach(() => {
    stageSelector(defaultSelector, {})
    wrapper = mount(<BUCTools {...initialMockProps} />)
  })

  afterEach(() => {
    wrapper.unmount()
  })

  it('Render: match snapshot', () => {
    expect(wrapper.isEmptyRender()).toBeFalsy()
    expect(wrapper).toMatchSnapshot()
  })

  it('Render: has proper HTML', () => {
    expect(wrapper.exists('[data-test-id=\'a-buc-c-buctools__panel-id\']')).toBeTruthy()
    expect(wrapper.exists('[data-test-id=\'a-buc-c-buctools__tabs-id\']')).toBeTruthy()
  })

  it('Render: has proper HTML: P5000 tab', () => {
    expect(wrapper.exists('[data-test-id=\'a-buc-c-buctools__P5000-button-id\']')).toBeTruthy()
    expect(wrapper.exists('[data-test-id=\'a-buc-c-buctools__tags-select-id\']')).toBeFalsy()
    expect(wrapper.exists('[data-test-id=\'a-buc-c-buctools__comment-textarea-id\']')).toBeFalsy()
    expect(wrapper.exists('[data-test-id=\'a-buc-c-buctools__comment-save-button-id\']')).toBeFalsy()
  })

  it('Render: has proper HTML: tag tab', () => {
    wrapper = mount(<BUCTools {...initialMockProps} initialTab={1} />)
    expect(wrapper.exists('[data-test-id=\'a-buc-c-buctools__P5000-button-id\']')).toBeFalsy()
    expect(wrapper.exists('[data-test-id=\'a-buc-c-buctools__tags-select-id\']')).toBeTruthy()
    expect(wrapper.exists('[data-test-id=\'a-buc-c-buctools__comment-textarea-id\']')).toBeFalsy()
    expect(wrapper.exists('[data-test-id=\'a-buc-c-buctools__comment-save-button-id\']')).toBeFalsy()
  })

  it('Render: has proper HTML: comment tab', () => {
    wrapper = mount(<BUCTools {...initialMockProps} initialTab={2} />)
    expect(wrapper.exists('[data-test-id=\'a-buc-c-buctools__P5000-button-id\']')).toBeFalsy()
    expect(wrapper.exists('[data-test-id=\'a-buc-c-buctools__tags-select-id\']')).toBeFalsy()
    expect(wrapper.exists('[data-test-id=\'a-buc-c-buctools__comment-textarea-id\']')).toBeTruthy()
    expect(wrapper.exists('[data-test-id=\'a-buc-c-buctools__comment-save-button-id\']')).toBeTruthy()
  })

  it('UseEffect: fetches tag list', () => {
    stageSelector(defaultSelector, { tagList: undefined })
    mount(<BUCTools {...initialMockProps} />)
    expect(getTagList).toHaveBeenCalled()
  })

  it('Handling: Changing tags', () => {
    (initialMockProps.onTagChange as jest.Mock).mockReset()
    wrapper = mount(<BUCTools {...initialMockProps} initialTab={1} />)
    const tagSelect = wrapper.find('[data-test-id=\'a-buc-c-buctools__tags-select-id\']').find('input').hostNodes()
    tagSelect.simulate('keyDown', { key: 'ArrowDown', keyCode: 40 })
    tagSelect.simulate('keyDown', { key: 'ArrowDown', keyCode: 40 })
    tagSelect.simulate('keyDown', { key: 'Enter', keyCode: 13 })
    expect(initialMockProps.onTagChange).toHaveBeenCalledWith([{
      label: 'buc:tag-refusjonskrav',
      value: 'tag-refusjonskrav'
    }, {
      label: 'buc:' + allTags[1],
      value: allTags[1]
    }])
  })

  it('Handling: Changing comments', () => {
    (saveBucsInfo as jest.Mock).mockReset()
    wrapper = mount(<BUCTools {...initialMockProps} initialTab={2} />)
    const newComment = 'this is a new comment'
    expect(wrapper.exists('[data-test-id=\'a-buc-c-buctools__comment-textarea-id\']')).toBeTruthy()
    expect(wrapper.find('[data-test-id=\'a-buc-c-buctools__comment-div-id\']').hostNodes().length).toEqual(bucInfo.comment!.length)

    expect(wrapper.find(TextArea).props().value).toEqual('')

    wrapper.find('[data-test-id=\'a-buc-c-buctools__comment-textarea-id\']').hostNodes().simulate('change', { target: { value: newComment } })
    wrapper.find('[data-test-id=\'a-buc-c-buctools__comment-save-button-id\']').hostNodes().simulate('click')

    expect(saveBucsInfo).toHaveBeenCalledWith(expect.objectContaining({
      comment: (bucInfo.comment as Comments)!.concat([{ value: newComment } as Comment])
    }))
  })

  it('Handling: Deleting comments', () => {
    (saveBucsInfo as jest.Mock).mockReset()
    jest.spyOn(global, 'confirm' as any).mockReturnValueOnce(true)
    wrapper = mount(<BUCTools {...initialMockProps} initialTab={2} />)
    expect(wrapper.find('[data-test-id=\'a-buc-c-buctools__comment-div-id\']').hostNodes().length).toEqual(bucInfo.comment!.length)

    expect(wrapper.find(TextArea).props().value).toEqual('')

    wrapper.find('[data-test-id=\'a-buc-c-buctools__comment-delete-0-id\']').hostNodes().simulate('click')

    expect(saveBucsInfo).toHaveBeenCalledWith(expect.objectContaining({
      comment: (bucInfo.comment as Comments)!.splice(1, 1)
    }))
  })

  it('Handling: Loads SEDs for P5000', () => {
    (initialMockProps.setMode as jest.Mock).mockReset()
    wrapper.find('[data-test-id=\'a-buc-c-buctools__P5000-button-id\']').hostNodes().simulate('click')
    expect(initialMockProps.setMode).toHaveBeenCalled()
  })
})
