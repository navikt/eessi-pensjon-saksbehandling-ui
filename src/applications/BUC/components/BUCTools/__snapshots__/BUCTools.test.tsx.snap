// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applications/BUC/components/BUCTools/BUCTools Render: match snapshot 1`] = `
.c1 {
  opacity: 0;
  -webkit-transform: translateX(20px);
  -ms-transform: translateX(20px);
  transform: translateX(20px);
  -webkit-animation: dgjvLb 0.3s forwards;
  animation: dgjvLb 0.3s forwards;
}

.c1.loading {
  background-color: rgba(128,128,128,0.2);
}

ui-fss.g72[id="c0"] {
  content: "c1,";
}

<div
  class="navds-panel c0 c1 navds-panel--border"
  data-testid="a_buc_c_buctools--panel-id"
>
  <div
    class="navds-tabs navds-tabs--medium"
    data-orientation="horizontal"
    data-testid="a_buc_c_buctools--tabs-id"
    dir="ltr"
  >
    <div
      class="navds-tabs__tablist-wrapper"
    >
      <div
        aria-orientation="horizontal"
        class="navds-tabs__tablist"
        data-orientation="horizontal"
        role="tablist"
        style="outline: none;"
        tabindex="0"
      >
        <button
          aria-controls="radix-:r0:-content-P5000"
          aria-selected="false"
          class="navds-tabs__tab navds-tabs__tab--medium navds-tabs__tab-icon--left"
          data-orientation="horizontal"
          data-radix-collection-item=""
          data-state="inactive"
          id="radix-:r0:-trigger-P5000"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class="navds-tabs__tab-inner navds-body-short"
          >
            buc:form-labelP5000
          </span>
        </button>
        <button
          aria-controls="radix-:r0:-content-P4000"
          aria-selected="false"
          class="navds-tabs__tab navds-tabs__tab--medium navds-tabs__tab-icon--left"
          data-orientation="horizontal"
          data-radix-collection-item=""
          data-state="inactive"
          id="radix-:r0:-trigger-P4000"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class="navds-tabs__tab-inner navds-body-short"
          >
            buc:form-labelP4000
          </span>
        </button>
        <button
          aria-controls="radix-:r0:-content-tags"
          aria-selected="false"
          class="navds-tabs__tab navds-tabs__tab--medium navds-tabs__tab-icon--left"
          data-orientation="horizontal"
          data-radix-collection-item=""
          data-state="inactive"
          id="radix-:r0:-trigger-tags"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class="navds-tabs__tab-inner navds-body-short"
          >
            ui:tags
          </span>
        </button>
      </div>
    </div>
    <div
      aria-labelledby="radix-:r0:-trigger-P5000"
      class="navds-tabs__tabpanel"
      data-orientation="horizontal"
      data-state="inactive"
      hidden=""
      id="radix-:r0:-content-P5000"
      role="tabpanel"
      tabindex="0"
    />
    <div
      aria-labelledby="radix-:r0:-trigger-P4000"
      class="navds-tabs__tabpanel"
      data-orientation="horizontal"
      data-state="inactive"
      hidden=""
      id="radix-:r0:-content-P4000"
      role="tabpanel"
      tabindex="0"
    />
    <div
      aria-labelledby="radix-:r0:-trigger-tags"
      class="navds-tabs__tabpanel"
      data-orientation="horizontal"
      data-state="inactive"
      hidden=""
      id="radix-:r0:-content-tags"
      role="tabpanel"
      tabindex="0"
    />
  </div>
</div>
`;
